From 5384a835a002935d43d0ad9f8c971ffa33c5a427 Mon Sep 17 00:00:00 2001
From: David Howells <dhowells@redhat.com>
Date: Tue, 25 Nov 2014 16:33:55 -0500
Subject: [fs] call rename2 if exists

Message-id: <20141125163355.18115.18867.stgit@warthog.procyon.org.uk>
Patchwork-id: 100914
O-Subject: [PATCH 14/47] fs: call rename2 if exists
Bugzilla: 985875
RH-Acked-by: Steve Best <sbest@redhat.com>

From: Miklos Szeredi <mszeredi@suse.cz>

Christoph Hellwig suggests:

1) make vfs_rename call ->rename2 if it exists instead of ->rename
2) switch all filesystems that you're adding NOREPLACE support for to
   use ->rename2
3) see how many ->rename instances we'll have left after a few
   iterations of 2.

Signed-off-by: Miklos Szeredi <mszeredi@suse.cz>
Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Original-commit: 7177a9c4b509eb357cc450256bc3cf39f1a1e639
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/namei.c b/fs/ext4/namei.c
index e3ac6fa..a5f24ae 100644
--- a/fs/ext4/namei.c
+++ b/fs/ext4/namei.c
@@ -3435,7 +3435,6 @@ const struct inode_operations ext4_dir_inode_operations = {
  .mkdir  = ext4_mkdir,
  .rmdir  = ext4_rmdir,
  .mknod  = ext4_mknod,
- .rename  = ext4_rename,
  .rename2 = ext4_rename2,
  .setattr = ext4_setattr,
  .setxattr = generic_setxattr,
diff --git a/fs/fuse/dir.c b/fs/fuse/dir.c
index 8f378c2..aa43bb6 100644
--- a/fs/fuse/dir.c
+++ b/fs/fuse/dir.c
@@ -848,12 +848,6 @@ static int fuse_rename2(struct inode *olddir, struct dentry *oldent,
  return err;
 }
 
-static int fuse_rename(struct inode *olddir, struct dentry *oldent,
-         struct inode *newdir, struct dentry *newent)
-{
- return fuse_rename2(olddir, oldent, newdir, newent, 0);
-}
-
 static int fuse_link(struct dentry *entry, struct inode *newdir,
        struct dentry *newent)
 {
@@ -1942,7 +1936,6 @@ static const struct inode_operations fuse_dir_inode_operations = {
  .symlink = fuse_symlink,
  .unlink  = fuse_unlink,
  .rmdir  = fuse_rmdir,
- .rename  = fuse_rename,
  .rename2 = fuse_rename2,
  .link  = fuse_link,
  .setattr = fuse_setattr,
diff --git a/fs/namei.c b/fs/namei.c
index 3d108a8..b1e1e40 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -4015,7 +4015,7 @@ int vfs_rename(struct inode *old_dir, struct dentry *old_dentry,
  if (error)
   return error;
 
- if (!old_dir->i_op->rename)
+ if (!old_dir->i_op->rename && !old_dir->i_op->rename2)
   return -EPERM;
 
  if (flags && !old_dir->i_op->rename2)
@@ -4074,10 +4074,11 @@ int vfs_rename(struct inode *old_dir, struct dentry *old_dentry,
   if (error)
    goto out;
  }
- if (!flags) {
+ if (!old_dir->i_op->rename2) {
   error = old_dir->i_op->rename(old_dir, old_dentry,
            new_dir, new_dentry);
  } else {
+  WARN_ON(old_dir->i_op->rename != NULL);
   error = old_dir->i_op->rename2(old_dir, old_dentry,
             new_dir, new_dentry, flags);
  }
-- 
1.7.1